### PR TITLE
Try: Reduce specificity of reset & classic styles.

### PR DESCRIPTION
**Pull request by jasmussen**
## Description

We have two stylesheets that are created to normalize the editing canvas: `reset.scss` and `classic.scss`.

- The first one exists because the editing canvas in the post editor is not inside an iframe, so it is subject to CSS bleed from wp-admin and other styles. The reset unsets a number of element styles to provide a blank canvas to build theme styles on.
- The other one, `classic.scss` is loaded for classic themes, and primarily serves to provide left/right/wide/full-wide alignment styles. Block themes do not need these, as those alignments are served by a `layout` property and handled by the system.

In both cases, while the styles are necessary, they should have as low specificity as is possible. The lower the specificity, the easier it is for a theme to style things in a nice way. 

There is a new selector, called `:where`, which is able to provide default styles with lowered specificity. Some of the details on how that works are outlined in [this article](https://css-tricks.com/almanac/selectors/w/where/), but you can also check out [this codepen for an example](https://codepen.io/joen/pen/abJVrGX). That selector is theoretically ideal for these resets, because it allows us to reduce the specificity quite a bit, letting both theme and block styles win any specificity fight they get into. In this PR, I tried employing that selector for those resets. 

Overall, it worked very well, especially for the reset which just feels perfect. Here's before and after using "empty theme" — and they should look identical, only with the new version having lower specificity:

Screenshot 2021-06-14 at 12 39 12

After:

Screenshot 2021-06-14 at 12 40 05

For other themes, here I tested TT1, it solved [this problem](https://github.com/WordPress/gutenberg/pull/32574#pullrequestreview-682298435) surfaced by @carolinan, where `margin-left: auto; margin-right: auto;` overrides the default navigation item margin, causing them to collapse. Before:

beftt1

(don't mind the padding, that's a separate issue)

After:

afttt1

_In these screenshots, the navigation menu is justified right_.

The primary "gotcha" I encountered is that themes commonly apply a `margin-left: 0;` on `ol` and `ul` elements, which seems an innocent and sensible change over the defaults. The problem is that for classic themes, the `margin-left: auto; margin-right: auto;` rules exist (along with a `max-width`) to center a _block_ inside a post content column that supports wide and fullwide images. When left margins are zeroed out, it effectively means a left-aligned list block. In trunk, the auto-margins win due to their intrinsic specificity. But with the lowered specificity of this branch, the zero margin now wins. 

For that reason, I included [a targeted](https://github.com/WordPress/gutenberg/compare/try/reduced-specificity-resets?expand=1#diff-5763ee427967f4474bd5a1cd169547cd483a1955b2b705e8c61b898da999f5cbR11) rule for `ol` and `ul` elements, to keep their specificity. It's not an inclusion I love, but it feels necessary as the change feels very common. What do you think?

## How has this been tested?

- Please test a variety of blocks and content in this branch. The demo content is good, but please at least include a justified-right navigation block, an image, and a list block. 
- Please test wide and fullwide blocks.
- Please test primarily in classic themes like TT1 or older.
- Please also test Empty Theme and verify that before and after are identical.

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. 
- [x] My code follows the accessibility standards. 
- [x] I've tested my changes with keyboard and screen readers. 
- [x] My code has proper inline documentation. 
- [x] I've included developer documentation if appropriate. 
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR (please manually search all `*.native.js` files for terms that need renaming or removal).

### CODE REVIEW

**Comment by jasmussen** [comment]
Note to self, need to push a fix for the Gallery block:

Screenshot 2021-06-14 at 12 57 09

This rule is now winning:

Screenshot 2021-06-14 at 12 57 04

Easy fix, but will let you all add some thoughts first.

**Comment by youknowriad** [comment]
> For that reason, I included a targeted rule for ol and ul elements, to keep their specificity. It's not an inclusion I love, but it feels necessary as the change feels very common. What do you think?

Interesting use-case, potentially we could consider doing this PR for just "resets" for now maybe? I think there might be other cases where themes could inadvertantly override the default centering margins (like figure maybe)

**Comment by jasmussen** [comment]
> Interesting use-case, potentially we could consider doing this PR for just "resets" for now maybe? I think there might be other cases where themes could inadvertantly override the default centering margins (like figure maybe)

Agreed, even if we don't land this for the classic styles, it seems a win to land it for the reset. That said, this PR does solve an issue with navigation block justification in classic themes, which isn't trivial to fix otherwise: if I increase the specificity in the navigation block, I break global styles.

**Comment by youknowriad** [comment]
🤔 What happens if we do this to the editor styles as well (the ones injected dynamically)?

**Comment by jasmussen** [comment]
Blowing my mind with that idea. Do you happen to recall where we rewrite `body` to `editor-styles-wrapper` ? Happy to give it a shot.

I wanted to also add the thought that we could make the changes suggested here to everything except the left/right auto margins. While that breaks the nav block issue I noted (I can look for other fixes), it would still let us lower the specificity of the top/bottom margin rules, which would be a nice win.

**Comment by youknowriad** [comment]
@jasmussen it happens here https://github.com/WordPress/gutenberg/blob/trunk/packages/block-editor/src/utils/transform-styles/transforms/wrap.js

basically, the ideal there would be to wrap without altering the specificity, not sure that's possible though

**Comment by jasmussen** [comment]
I gave it a shot, but I think that lowers the wrong specificity:

Screenshot 2021-06-14 at 13 15 30

So this margin rule from the theme:

Screenshot 2021-06-14 at 13 15 46

Now it loses to the block style:

Screenshot 2021-06-14 at 13 15 37

In other words, it does work as intended. But we need to apply this to everything we _want_ to be overridden. So in the case of the left aligned gallery, I can add the `where` to the editor style and it'd work. 

> basically, the ideal there would be to wrap without altering the specificity, not sure that's possible though

We could potentially use `:is` here: https://css-tricks.com/almanac/selectors/i/is/#specificity-of-is — but I may be missing some nuance of what you're saying.

**Comment by gwwar** [comment]
> There is a new selector, called :where, which is able to provide default styles with lowered specificity. 

TIL, that's pretty neat! ✨ 

Zooming out a bit, did we have specificity targets for different parts of the editor? Using a comma delimited notation (style, id, class, element) with resets we'd probably stick with might have scores from (0,0,0,X), while blocks might be aiming for something like (0,0,1,X) to (0,0,2,X)?

As a reminder for folks 👀 CSS specificity is mostly counting. And no matter the number of items in one category it can't beat the next one. Using a comma delimited notation, if I say had 15 element rules in a selector a single class selector would still beat it.

In Joen's example in https://github.com/WordPress/gutenberg/pull/32659#issuecomment-860605869

This would be (0,0,1,1)
121884045-c1d50280-cd12-11eb-97c9-e9adb084008f

vs (0,0,1,0) (with the :where pseudo selector).
121884026-bc77b800-cd12-11eb-96d3-f19adde3e0eb

**Comment by jasmussen** [comment]
> Zooming out a bit, did we have specificity targets for different parts of the editor?

It's been adhoc so far, but it might be good to start thinking about this in a more systematic way, in part because we have this new `:where` tool to provide very soft defaults, but perhaps more importantly because for global styles to work, the specificity of block styles has to be as low as possible for any theme.json specified overrides to take hold. 

Do you have any good suggestions on how to start systematizing that? Are there any principles we can extract in documentation? Any linters we can add?

**Comment by jasmussen** [comment]
Curious, though, how did you calculate the specificity of `.editor-styles-wrapper :where([data-block])`? I suspect your value is correct, but I wouldn't know how to parse that particularly gnarly selector into the three categories, and online calculators get it wrong, probably due to unawareness of the particular qualities of `:where`; [this one](https://specificity.keegan.st/) gives me 0,3,0.

**Comment by gwwar** [comment]
> Curious, though, how did you calculate the specificity of .editor-styles-wrapper :where([data-block])?

Manually counting, checkout also https://www.w3.org/TR/selectors/#specificity. 

From the spec [:where and it's argument are always zero](https://www.w3.org/TR/selectors/#zero-matches)

> The Specificity-adjustment pseudo-class, :where() [...] neither the **:where() pseudo-class, nor any of its arguments**, contribute to the specificity of the selector—**its specificity is always zero**.

So with that in mind `.editor-styles-wrapper :where([data-block])` simplifies to (0,0,1,0). And we can see that `figure.wp-block-gallery` does indeed beat it, which makes sense since it has a specificity of (0,0,1,1).

> Do you have any good suggestions on how to start systematizing that? Are there any principles we can extract in documentation? Any linters we can add?

So probably the hierarchy of styling will probably be something like:

Reset / Base Styles -> Block Styles -> [ Custom Theme / Editor Styles ] -> Global Styles -> Specific Block User Overrides

If we can provide consistent specificity targets at the first two layers, then it should be easier to provide guidelines for folks building themes, and the necessary overrides for anything above that. Note that we don't necessarily have to use strict specificity guidelines if we're sandboxing styling in other ways (say isolating rules via iframes and such).

**Comment by jasmussen** [comment]
Thank you as always for teaching me, I very much appreciate that.

Your hierarchy reminds me of https://github.com/WordPress/gutenberg/issues/20331, which seems like healthy serendipity worth building on. The tricky bit is the theme, as it requires some care on part of the author to not mess up the global style user overrides. However as those same properties are provided by theme.json directly, hopefully that helps manage the complexity.

**Comment by jasmussen** [comment]
I've got two navigation block PRs that are blocked on this one landing first: #33021 and #32574. I'd really love a review of this one, and ideally a green check because all it really does it reduce specificity so it should be safe. Thanks.

**Comment by youknowriad** [comment]
> Reset / Base Styles -> Block Styles -> [ Custom Theme / Editor Styles ] -> Global Styles -> Specific Block User Overrides

This hierarchy makes sense but let me clarify the difficulty a bit here. In the frontend, the equivalent hierarchy is:

```
Block Styles ->  Custom Theme -> Global Styles -> Specific Block User Overrides
```

So our goal is to match both hierarchies as closely as possible, there are thought two main differences:

1- in the editor we need a `Reset / Base Styles`, this reset styles is basically something like:

```
.editor-styles-wrapper p { all: revert; }
```

that's just an example, the important bit here is that we're trying to reset the styles of the base elements to the browser's defaults by overriding the WP-Admin common.css styles. But by doing so we increased the specificity of `p` which means this creates a potential issue with the styles that are loaded later "Block Styles" and "Theme Editor Styles".

2- To solve the issue above what we did (a long time ago) is to add `.editor-styles-wrapper` to all editor styles. Meaning a theme writing `p { color: red }` in its editor styles will result in the following being loaded:

```
.editor-styles-wrapper p { color: red; }
```

---
The problem remains though for `Block Styles` which we don't wrap in `.editor-styles-wrapper` yet. So we have two potential solutions:

 1- Try to wrap all the styles that can affect blocks with `.editor-styles-wrapper`, the important remaining one for us is "Block Styles" (the block-library package stylesheets)

 2- Make sure the "reset" and the editor styles don't affect the specificity of their rules when they wrap the CSS with. `.editor-styles-wrapper` which this PR is attempting (I think), I think we're not there yet though.

**Comment by jasmussen** [comment]
Thanks @youknowriad and @gwwar for the reviews. Before we go deeper on the path of a hierarchy, is this PR okay to land as an interim step?

**Comment by youknowriad** [comment]
I think this looks good to me because the "reset / base styles" is the first item in the hierarchy, meaning if we reduce its specificity, it can only have positive result (and no negative one) but I feel like we should be consistent about the way we wrap our styles, meaning if we apply this principle `:where` for the "reset" rules, we should look at applying the same ones to "custom theme editor editor styles" and maybe even try to find a way to do it for "block library styles" (though this one we never did, so not sure how exactly).

**Comment by youknowriad** [comment]
I'd love to get @gwwar's input on this as well :)

**Comment by gwwar** [comment]
Yes, I do think `:where` makes sense for resets and base styles, so our use case here makes a lot of sense. (Future specification might also let us explicitly set the specificity we want in the :where psuedo class).

I'll try to manually test this one today

**Comment by gwwar** [comment]
For folks reviewing, the empty theme is available at https://github.com/WordPress/theme-experiments/tree/master/emptytheme

One way to test is to clone the theme-experiments repo, then symlink the emptytheme to your local test instance's `/wp-content/themes/emptytheme`. Props to @MaggieCabrera for pointing out where this lives ✨ 

It'll look a bit like:
```bash
git clone git@github.com:WordPress/theme-experiments.git
cd theme-experiments
ln -s /your-git-checkout-path/theme-experiments/emptytheme  /your-path-to-a-local-wordpress-instance/wp-content/themes/emptytheme
```

**Review by gwwar — APPROVED** [review]
Thanks for looking into this @jasmussen I'm going to give tentative approval here. Things look pretty safe, but it's pretty likely we might have missed a few unintended side effects. It might be worth giving a folks a heads up after merging this one into trunk.

One thing I did notice was that selecting a justified navigation block would trigger a left-align shift. I don't think this happens in trunk. I'd maybe rebase to see if that clears it up.

https://user-images.githubusercontent.com/1270189/124800314-5b569500-df0a-11eb-9c84-f4367f0ae469.mp4

Otherwise I think things mostly look the same:

| before | after |
|-----|-----|
| Screen Shot 2021-07-07 at 10 00 10 AM | after2 | 

| before | after |
|-----|-----|
| Screen Shot 2021-07-07 at 10 01 07 AM | after |

**Comment by jasmussen** [comment]
Thank you for the review, your reviews are the best, and excellent catch with the onselect alignment. Here's trunk:

trunk

Here's this branch before rebase:

this branch

Here's the branch after rebase:

after rebase

Two immediate take-aways: you were completely right about the onselect alignment thing. I actually had a separate fix for that in #32574, but that may no longer be necessary, since it's possible it was inadvertently fixed by #33088 (which intended to be a more general refactor).

The other take-away is the added margin and padding. This is a side-effect of the reduced specificity of these styles, you can see in this GIF how the added margin and padding rules are from TT1:

tt1

Those rules are rather old and were wrong even before this PR, I've trac'ed them here: https://core.trac.wordpress.org/ticket/53220

There might be a few other such issues caused by theme styles having previously being overridden by higher specificity, now winning out. But I still think it's worth it to land this PR and "de-escalate" the specificity wars.

Thanks again for the review, I'll ask for a sanity check again before I merge.

**Comment by Mamaduka** [comment]
In general, I think this looks good. I found few minor issues that can be addressed in follow-up PRs.

Tested with two WP installations side by side, using Demo content + Navigation (horizontal) justified to the right.

### Installations:
A. Gutenberg build with this branch
B. Gutenberg 11.0

### Themes tested
* TT1
* TT1 Blocks
* 2020
* 2019

### Minor issues I spotted on this branch
**TT1 Blocks**
* None

**TT1**
* The gallery with no alignment isn’t centered.

**2020**
* Quote with no alignment isn’t centered.
* Gallery with no alignment isn’t centered.

**2019**
 * Lists are centered.

### Screenshots


| this branch  | 11.0 |
| ------------- | ------------- |
| CleanShot 2021-07-08 at 13 14 58  | CleanShot 2021-07-08 at 13 15 08 |

| this branch  | 11.0 |
| ------------- | ------------- |
| CleanShot 2021-07-08 at 13 21 38  | CleanShot 2021-07-08 at 13 21 55  |

**Comment by jasmussen** [comment]
Thanks for looking. I have this feeling those are all issues best addressed in the theme itself, does that seem right to you? If so, I'd be happy to create tickets for those and/or look at patches. Or did you get a sense that this was fixable in the block editor, without raising specificity again?

**Comment by Mamaduka** [comment]
I'm not sure. I think some should be fixed in Gutenberg and some in themes.

In the case of the Gallery with no alignment, `margin: 0` is coming from block's `editor.css` - https://github.com/WordPress/gutenberg/blob/trunk/packages/block-library/src/gallery/editor.scss#L22.

2020 theme adds the `margin: 0` rule for blockquote in the editor, but then blockquotes are correctly centered on the front-end.

**Comment by jasmussen** [comment]
Excellent catch on the gallery. It appears to not be a problem with block themes as the left/right auto margins are enforced as !important. 

But it is indeed a problem in classic themes, as the margin: 0 rule now wins out:

Screenshot 2021-07-08 at 12 59 08

While I still think lowering that specificity, and actually updating the rule of the Gallery block to not have such a blanket zero margin output, it does make me question whether it's actually a good idea to update those classic rules at all. We might want to do it for the reset styles only after all 🤔 

Block themes will still benefit from the new lowered specificity of the reset style. What do you think? And thanks again for looking.

**Comment by jasmussen** [comment]
Though it is worth noting, that if we _don't_ include the lowered specificity, then improvements from https://github.com/WordPress/gutenberg/pull/33021, just as an example, can't benefit classic themes. This will likely expand more widely as more properties like margins and paddings are absorbed into global styles.

**Comment by Mamaduka** [comment]
The Gallery block issue would most visible, so it might be good to fix that as a follow-up.

I think it would be easier to gather more feedback if we merged this PR.

**Comment by jasmussen** [comment]
I'm waffling on merging it, as I'm going on vacation next week, so I won't have time to follow up on this. So at this moment I'm leaning towards pausing this one until I get back.

**Comment by jasmussen** [comment]
I'm back from AFK and hope to land this soon. Since it also appears to improve [a 2019 theme issue](https://core.trac.wordpress.org/ticket/53872), it would be good to land soon. I've rebased to test that it's still valid, and will be investigating the gallery block issue and how that can be fixed, and then either push a fix to this branch, or have a followup ready.

**Comment by jasmussen** [comment]
Alright, after investigating a bit further, it appears that the issue, as so well reviewed by @Mamaduka in https://github.com/WordPress/gutenberg/pull/32659#issuecomment-876292841, is primarily an issue in classic themes. Here's what happens in 2020 that causes the gallery to be left aligned. The theme outputs this rule:

Screenshot 2021-08-05 at 11 35 10

This rule is common for themes, as the `figure` element is born with intrinsic margins, including left and right margins. But with the new lowered specificity for the left/right auto margins, the following rule that is intended to center the top level block now loses out in specificity:

Screenshot 2021-08-05 at 11 35 00

So there are two things we can do here.

**1. Restore specificity for those auto margins.** 

This would be a bit of a bummer as I recall those margins being frustrating when I mostly worked on classic themes. But it might be a pragmatic choice for backward compatibility. Block themes won't have to worry about them, at least. Keep something like this in classic.scss, in other words:

```
// Provide baseline auto margin for centering blocks.
// Specificity is kept at this level as many classic themes output
// rules like figure { margin: 0; } which would break centering.
.editor-styles-wrapper .wp-block {
	margin-left: auto;
	margin-right: auto;
}
```

**2. Expand the list of blocks with specific auto margins.**

This PR already had the following in place, to center lists:

```
// This needs specificity to override theme editor styles
// such as ol { margin-left: 0; }.
.editor-styles-wrapper ol.wp-block,
.editor-styles-wrapper ul.wp-block {
	margin-left: auto;
	margin-right: auto;
}
```

We could expand that definition to include the figure element, which fixes the 2020 issue and still keeps the generic .wp-block class low-specificity. Something like this:

```
// This needs specificity to override theme editor styles
// such as ol { margin-left: 0; }.
.editor-styles-wrapper figure.wp-block,
.editor-styles-wrapper ol.wp-block,
.editor-styles-wrapper ul.wp-block {
	margin-left: auto;
	margin-right: auto;
}
```

It fixes it for 2020:

Screenshot 2021-08-05 at 11 47 07

The downside to that approach is that it very slightly _increases_ the specificity of those margin rules for classic themes. 0,2,1 for `.editor-styles-wrapper figure.wp-block` compared to 0,2,0 for `.editor-styles-wrapper .wp-block`.

Personally I'm slightly in favor of option 2, because those margins are annoying. But it's not a strong preference, and I'd be happy to go the potentially more pragmatic route.

Before I push anything, I'd love your thoughts. Thanks.

**Comment by youknowriad** [comment]
> 2. Expand the list of blocks with specific auto margins.

Personally I think this is a mistake and should be removed from all blocks, it shouldn't be the responsibility of the block to define its margins but should be the responsibility of its container ideally. Of course there might be special cases to clear out margins... but margin auto is very specific to the "flow" layout.

**Comment by jasmussen** [comment]
Going forward with block themes, there won't be any auto margin applied. But given how legacy themes have assumed those auto margins for a while, would you be okay with option 1 then?

**Comment by youknowriad** [comment]
Yes, I think option 1 is better. These "classic" styles are not "reset" styles meaning It's not clear whether they should have less specificity or more than editor styles. So I think keeping that rule as is is a good approach.

**Comment by jasmussen** [comment]
Awesome. Things still work well here:

Screenshot 2021-08-05 at 12 17 09

File: packages/edit-post/src/classic.scss
Comment (youknowriad) [inline packages/edit-post/src/classic.scss:14]: Noting that these margins can also get removed by `figure { margin: 0; }` do you think we should also move them up?

Comment (jasmussen) [reply packages/edit-post/src/classic.scss:14]: I'd personally prefer to keep them there, as they as I recall were the most disruptive to work with, and the reduced specificity of those margins were the biggest win of this approach. It should also be less disruptive as it's vertical margins.

However I'm happy to move them up either in this PR or as a followup if it turns out to be necessary?

Comment (youknowriad) [reply packages/edit-post/src/classic.scss:14]: >  the biggest win of this approach.

I'm curious, in which situation the smaller specificity benefit classic themes here. What kind of styles classic themes apply here?

Comment (jasmussen) [reply packages/edit-post/src/classic.scss:14]: I'm thinking mostly of developing new classic themes, or refactoring older ones.

Comment (kjellr) [reply packages/edit-post/src/classic.scss:14]: I agree with @jasmussen that it should be fine to keep these where they are. I tested in a few themes and didn't notice this causing any issues. If they did interfere with that `figure` rule, it'll just be the vertical margin (and may actually have no effect if there's margin on the items above + below the `figure` anyway).

Comment (jasmussen) [reply packages/edit-post/src/classic.scss:14]: Thanks for looking Kjell. Cool, in any case it's an easy follow up which I will be keeping my eyes peeled on and ready to change if need be. I'd like to land this PR tomorrow or Monday, so there's still time to chime in here.

**Comment by jasmussen** [comment]
Alright, it's Monday and I'll be here for any followups. This one after recent iterations feels very safe to me, so I expect when I merge it in a moment it will be a non issue. Nevertheless, let's keep an eye out on theme styles in the editing canvas!

**Pull request merged**
